src/inmanta/module.py

```diff
@@ -1144,21 +1243,36 @@ def rewrite_version(cls: Type[TModuleMetadata], source: str, new_version: str) -
         if current_version == new_version:
             LOGGER.debug("Current version is the same as the new version: %s", current_version)
 
-        result: str = cls._substitute_version(source, new_version)
+        result: str = cls._substitute_version(source, new_version, version_tag)
 
         try:
             new_metadata = cls.parse(result)
         except Exception:
             raise Exception("Unable to rewrite module definition.")
 
-        if new_metadata.version != new_version:
-            raise Exception(f"Unable to write module definition, should be {new_version} got {new_metadata.version} instead.")
+        # Validate whether version field was updated correctly in metadata file
+        if cls is ModuleV2Metadata:
+            expected_version = new_version
+        else:
+            expected_version = f"{new_version}.{version_tag.lstrip('.')}" if version_tag else new_version
+        if new_metadata.version != expected_version:
+            raise Exception(
+                f"Unable to write version in module definition, should be {expected_version} "
+                f"got {new_metadata.version} instead."
+            )
+
+        # Validate whether version_tag field was updated correctly in metadata file
+        if isinstance(new_metadata, ModuleV2Metadata) and version_tag != new_metadata.version_tag:
+            raise Exception(
+                f"Unable to write tag_build in module definition, should be '{version_tag}' got "
+                f"'{new_metadata.version_tag}' instead."
+            )
```

Comment:
I see your point, but I am still on doubt whether this is an actual improvement, because this changes makes the error reporting more generic and as such also more vague. The location where the version tag is stored is dependent on the type of module (V1 vs V2). The previous error message pointed to the exact field that was causing issues, with the new implementation that's not possible. So you have to be very much aware about the differences between both types of modules to be able to interpret the error message. In our team that won't be an issue, but for external people it might be. Maybe I have to try to embed this information in the error message. I will give that a try.

Reply:
You make a good point. Alternatively we could offer an abstract method on the ABC to get the version and tag in a tuple, then stick with your logic but with polymorphism rather than isinstance?